drivers/dma/dma_stm32.c

```diff
@@ -0,0 +1,869 @@
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_SINGLE);	\
+		break;							\
+	case 4:								\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC4);	\
+		break;							\
+	case 8:								\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC8);	\
+		break;							\
+	case 16:							\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC16);	\
+		break;							\
+	default:							\
+		LOG_ERR("Burst size error");				\
+		return -ENOTSUP;					\
+} while (0)
+#endif
+
+static u32_t dma_stm32_width_config(struct dma_config *config,
+				    bool source_periph,
+				    DMA_TypeDef *dma,
+				    u32_t id)
+{
+	u32_t periph, memory;
+	u32_t m_size = 0, p_size = 0;
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+	u32_t periph_burst, memory_burst;
+#endif
+
+	if (source_periph) {
+		periph = config->source_data_size;
+		memory = config->dest_data_size;
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+		periph_burst = config->source_burst_length;
+		memory_burst = config->dest_burst_length;
+#endif
+	} else {
+		periph = config->dest_data_size;
+		memory = config->source_data_size;
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+		periph_burst = config->dest_burst_length;
+		memory_burst = config->source_burst_length;
+#endif
+	}
+	int index = find_lsb_set(config->source_data_size) - 1;
+
+	m_size = table_m_size[index];
+	index = find_lsb_set(config->dest_data_size) - 1;
+	p_size = table_p_size[index];
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+	switch (memory_burst) {
+		BURST_SIZE(Memory, M);
+	}
+	switch (periph_burst) {
+		BURST_SIZE(Periph, P);
+	}
+#endif
+
+	return (m_size | p_size);
+}
+
+static int dma_stm32_config_devcpy(struct device *dev, u32_t id,
+				   struct dma_config *config)
+
+{
+	enum dma_channel_direction direction = config->channel_direction;
+	const struct dma_stm32_config *dma_stm32_config =
+					dev->config->config_info;
+	DMA_TypeDef *dma = (DMA_TypeDef *)dma_stm32_config->base;
+	u32_t size_config;
+
+	switch (direction) {
+	case MEMORY_TO_PERIPHERAL:
+		size_config = dma_stm32_width_config(config, false, dma, id);
+		LL_DMA_ConfigTransfer(dma, table_stream[id],
+				LL_DMA_DIRECTION_MEMORY_TO_PERIPH |
+				LL_DMA_MODE_CIRCULAR |
+				LL_DMA_PERIPH_NOINCREMENT |
+				LL_DMA_MEMORY_INCREMENT |
+				size_config |
+				LL_DMA_PRIORITY_LOW);
+		break;
+	case PERIPHERAL_TO_MEMORY:
+		size_config = dma_stm32_width_config(config, true, dma, id);
+		LL_DMA_ConfigTransfer(dma, table_stream[id],
+				LL_DMA_DIRECTION_PERIPH_TO_MEMORY |
+				LL_DMA_MODE_CIRCULAR |
```

Comment:
I had to remove CIRCULAR here, otherwise it was not working.
I guess that in this particular case (PERIPHERAL_TO_MEMORY) it does not apply.
Maybe.

Reply:
Zephyr currently does not support to configure i DMA is in circular mode. I have not experienced with many DMA peripherals so I don't know if other DMA peripherals also have this feature. Before this feature is added, I'm thinking about configuring if it's in circular mode in dts. What do you think?

Reply:
I don't know what @erwango has to say here. But yes, it seems to me that all the dma
configurations sooner or later has to be put in dts. So we may start with circular mode.

Reply:
Here I have some doubts. We have two specific dts parts:

1. The dma controller section, which should describe the controller capabilities
2. The dma section inside a client driver (like I2S). Here we should describe what the client
    is using (channels, slot, direction,...)

Probably the CIRCULAR info goes here somehow.

Reply:
Yes, back to Linux we have DMA controller part dts and DMA device part dts, and Zephyr supports this arch for SPI and I2C already, I think I can try to add a dma-devices.yaml to add support for this function.

Reply:
Yes, seems the right direction

Reply:
I'm working on the dma consumer in #15218 , But I have some doubts, some properties like if this channel is working in circular mode is not a generic property for all dma drivers. Only DMA of stm32 supports this functionality. I don't think we should put it in dts, but configure it in the controller side. There would be a property to specify which channel is in circular mode, like this:

dma: dma@40012000 {
...
        circular-channel = <0 2 3>;
...
}

Just because it's special I think, and dma consumer currently is not able to handle device related special properties. What do you think?

Reply:
I just realized that this way low level dts's circular-channel would mistakely overwrite high level dts's circular-channel some time.

Reply:
@cybertale Can you add CIRCULAR support and make it configurable? In my implementation #14916  for UART Async, the DMA is configured to use the circular mode ( which I extended).

Reply:
Sure, will do it as soon as I got some time.